src/engine/renderer/glsl_source/reliefMapping_fp.glsl

```diff
@@ -26,37 +26,16 @@ uniform sampler2D	u_NormalMap;
 uniform int u_HeightMapInNormalMap;
 #endif // r_normalMapping || USE_PARALLAX_MAPPING
 
+#if defined(r_normalMapping)
+// HACK: fourth component tells renderer if normal scale is customized or not
```

Comment:
Why not simply put in (1, 1, 1) in that case?

Reply:
The true reliable test is to test of something is not zero, so everytime we allocate, we already know the default value is right. Otherwise, we have to check all the code to set the default value everytime we allocate, which is calling for future bugs in the future as things would be modified.

Note that patching this code is an iterative process. As second commit moves the test out of the glsl shader (doing it per-material instead of per-pixel), there is no need to hack a fourth component and a bool is used instead.

We maybe would be able to not use a bool at all by hacking the third channel (we know 0 is wrong) but that's something I'm not ready to support and it would add no performance improvement. The purpose of hacking a channel is to avoid to bind an additional variable from CPU to GPU. The second commit entirely remove the need for an additional variable so the bool costs nothing GPU side and costs no less CPU side.

Reply:
Unfortunately it would be complicated to remove the fourth channel hack in first commit without squashing second commit. And I would prefer to keep the two commits separated as it makes better to understand the iterative process involved in rewriting that code. This is done step by step and sometime some code has to be written that would be removed by the next step, this to avoid huge commits that does everything at once.

Reply:
Testing the z component in GLSL still has a small extra cost in terms of running extra instructions, compared to just setting 1,1,1. The condition has to be checked for every texture in every frame, instead of once at shader load time. Probably not worth worrying about though.